src/storage/mod.rs

```diff
@@ -102,34 +110,40 @@ where
     /// with mafintosh).
     /// TODO: Ensure the signature size is correct.
     /// NOTE: Should we create a `Data` entry type?
-    pub fn put_data(&mut self, index: usize, data: &[u8], nodes: &[Node]) -> Result<()> {
+    pub fn put_data(&mut self, index: u64, data: &[u8], nodes: &[Node]) -> Result<()> {
         if data.is_empty() {
             return Ok(());
         }
 
-        let range = self.data_offset(index, nodes)?;
+        let mut range = self.data_offset(index, nodes)?;
 
         ensure!(
-            range.len() == data.len(),
-            format!("length  `{:?} != {:?}`", range.len(), data.len())
+            range.by_ref().count() == data.len(),
+            format!("length  `{:?} != {:?}`", range.count(), data.len())
         );
 
-        self.data.write(range.start, data)
+        self.data.write(range.start, data).map_err(|e| anyhow!(e))
```

Comment:
Using `?` will map the error for you.

Reply:
Now I remember why I've done this `.map_err`.

My previous attempt to use `?` led me to unsized types:

```
error[E0277]: the size for values of type `dyn std::error::Error + std::marker::Send + std::marker::Sync` cannot be known at compilation time
  --> src\storage\mod.rs:80:59
   |
80 |         instance.bitfield.write(0, &header.to_vec()).await?;
   |                                                           ^ doesn't have a size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `dyn std::error::Error + std::marker::Send + std::marker::Sync`
   = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
   = note: required because of the requirements on the impl of `std::error::Error` for `std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>`
   = note: required because of the requirements on the impl of `std::convert::From<std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>>` for `anyhow::Error`
   = note: required by `std::convert::From::from`
```

It seems that `anyhow::Error` does not implement `std::error::Error` directly. This issue https://github.com/dtolnay/anyhow/issues/63 led me to attempt calling `.into()`, which also failed:

```
error[E0284]: type annotations needed
  --> src\storage\mod.rs:80:60
   |
80 |         instance.bitfield.write(0, &header.to_vec()).await.into()?;
   |         ---------------------------------------------------^^^^--
   |         |                                                  |
   |         |                                                  cannot infer type
   |         this method call resolves to `T`
   |
   = note: cannot resolve `<_ as std::ops::Try>::Ok == _`
```

Which also didn't work even with a type specification:

```
error[E0277]: the trait bound `std::result::Result<(), anyhow::Error>: std::convert::From<std::result::Result<(), std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>>>` is not satisfied
  --> src\storage\mod.rs:83:93
   |
83 |         let result: anyhow::Result<()> = instance.bitfield.write(0, &header.to_vec()).await.into();
   |                                                                                             ^^^^ the trait `std::convert::From<std::result::Result<(), std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>>>` is not implemented for `std::result::Result<(), anyhow::Error>`
   |
   = note: required because of the requirements on the impl of `std::convert::Into<std::result::Result<(), anyhow::Error>>` for `std::result::Result<(), std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>>`
```

There is an upstream issue documenting this conversion error https://github.com/dtolnay/anyhow/issues/66, but I'm not sure if it will fixed, as there is a conflict on the impl of the conversions as documented on the previous discussion.

That led me to `.map_err` and it compiled without needing to implement a local structure to wrap stderr.